### ISSUE TITLE
HMR - `WeakSet key must be an object` while exporting HostObject from file

### ISSUE THREAD

**Issue by jpudysz** [expo/expo#33670]
### Summary

Hello Expo team,  

I received a report from one of my users who was trying to use Unistyles 3 and encountered a weird error during hot reloading.  

Here’s a preview:  

https://github.com/user-attachments/assets/f7b5f803-4905-4b94-8c9a-1e910593e805

Here’s the full stack trace:  

![image](https://github.com/user-attachments/assets/3703cce4-da3e-4711-87b9-dac37e07b253)  

### Issue  

The developer is attempting to export `styles` from Unistyle's `StyleSheet.create`, which is a JSI HostObject.  
Trying to modify the style results in the following error, pointing to the HRM client and Metro.  

### Observations  

- There is no need to import `styles` anywhere. Using the `export` keyword already breaks it
- Changes made outside of `StyleSheet` that trigger hot reload also cause the same issue
- I verified that this doesn’t happen in a bare React Native project
- I’m not sure if the issue exists in the Old Architecture, as Unistyles 3 cannot run without Fabric
- I can also reproduce it on Android

### My Findings  

I temporarily modified the JSI code for `StyleSheet.create` and found that the issue seems to be related to returning a C++ HostObject.  When I returned an empty JSI object, no error was reported.  

Also it looks like that `JSI HostObject` can't be a key of `WeakSet`. 

### What platform(s) does this occur on?

Android, iOS

### SDK Version

52.0.18

### Environment

```text
expo-env-info 1.2.1 environment info:
    System:
      OS: macOS 15.0
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.15.0 - /usr/local/bin/node
      Yarn: 1.22.22 - /opt/homebrew/bin/yarn
      npm: 10.7.0 - /usr/local/bin/npm
      Watchman: 2024.11.25.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.16.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.0, iOS 18.0, macOS 15.0, tvOS 18.0, visionOS 2.0, watchOS 11.0
      Android SDK:
        API Levels: 23, 28, 29, 30, 33, 34, 35
        Build Tools: 30.0.3, 34.0.0, 35.0.0
        System Images: android-23 | ARM 64 v8a, android-28 | ARM 64 v8a, android-29 | ARM 64 v8a, android-30 | ARM 64 v8a, android-30 | Google APIs ARM 64 v8a, android-34 | Android TV ARM 64 v8a, android-34 | ARM 64 v8a, android-35 | Google APIs ARM 64 v8a
    IDEs:
      Android Studio: 2024.2 AI-242.23339.11.2421.12483815
      Xcode: 16.0/16A242d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~52.0.18 => 52.0.18 
      expo-router: 4.0.11 => 4.0.11 
      react: 18.3.1 => 18.3.1 
      react-dom: 18.3.1 => 18.3.1 
      react-native: 0.76.5 => 0.76.5 
      react-native-web: ~0.19.13 => 0.19.13 
    npmGlobalPackages:
      eas-cli: 11.0.2
    Expo Workflow: bare
```

### Minimal reproducible example

1. Download repo: https://github.com/jpudysz/hmr-expo-repro
2. Install dependencies
3. Go to `components/CText.tsx`
4. Change anything to cause Hot reload
5. Observe error

6. Remove export keyword
7. Change anything to cause Hot reload
8. Observe clean console

**Comment by brentvatne** [comment]
hi @jpudysz! is this only happening when a project uses expo-modules-core?

**Comment by jpudysz** [comment]
Yeah, for Bare RN everything works great. For projects created by expo cli it doesnt.

**Comment by chrfalch** [comment]
Hi, @jpudysz - I'm trying to look into this to see if I can find something obvious. After some digging I found that the problem seems to be an issue in `expo-router` and originates from `expo-router`'s fast-refresh hook where it looks at the exports in a module and uses a `WeakSet` to store exports - which fails when trying to use a `JSI::HostObject` as the WeakSet key. That's why removing the `export` keyword fixes the problem :)

I'll try to make a PR that fixes the issue.

**Comment by jpudysz** [comment]
Thanks for the update @chrfalch ! 👏
